File: webapp/storyteller/src/app/service/comment.service.ts
import { Injectable } from '@angular/core';
import {HttpClient} from "@angular/common/http";
import {environment} from "../../environments/environment";
import {AddCommentRequest, Comment} from "../model/comment";

@Injectable({
  providedIn: 'root'
})

/**
 * Service for Comments
 */
export class CommentService {

  constructor(private http: HttpClient) { }

  /**
   * Get all comments by their chapter id
   * @param chapterId the id of the chapter to get the comments from
   */
  getCommentsByChapterId(chapterId: string) {
    return this.http.get<Comment[]>(`${environment.apiUrl}/comments/${chapterId}`);
  }

  /**
   * Add a new comment
   * @param comment the comment to add
   */
  addComment(comment: AddCommentRequest) {
    return this.http.post<Comment>(`${environment.apiUrl}/comments/add`, comment);
  }

  /**
   * Get all comments by their block id and chapter id
   * @param chapterId the id of the chapter to get the comments from
   * @param blockId the id of the block to get the comments from
   */
  getCommentsByBlockId(chapterId: string, blockId: string) {
    return this.http.get<Comment[]>(`${environment.apiUrl}/comments/${chapterId}/${blockId}`);
  }
}

File: api/src/main/java/de/storyteller/api/service/comment/CommentService.java
package de.storyteller.api.service.comment;

import de.storyteller.api.v1.dto.comment.AddCommentRequest;
import de.storyteller.api.v1.dto.comment.CommentDTO;

import java.util.List;

/**
 * Service for handling comments
 */
public interface CommentService {
    /**
     * Add a comment to a chapter or block
     * @param addCommentRequest the request containing the comment
     * @return the added comment
     */
    CommentDTO addComment(AddCommentRequest addCommentRequest);

    /**
     * Get all comments for a chapter
     * @param chapterId the id of the chapter
     * @return the comments
     */
    List<CommentDTO> getCommentsByChapter(String chapterId);

    /**
     * Get all comments for a block
     * @param chapterId the id of the chapter
     * @param blockId the id of the block
     * @return the comments for a block
     */
    List<CommentDTO> getCommentsByBlockId(String chapterId, String blockId);
}
